importccl: support option to compress output files using gzip

C0rWin opened the pull request:
Fix #45579

This commit extends EXPORT functionality by enabling compression of the
exported stream as suggested in #45579. Currently only gzip is supported
and the export clause to use compression looks as following:

```
export into csv 's3://export.csv' with compression = gzip from select * from foo;
```

Signed-off-by: Artem Barger <bartem@il.ibm.com>

Release note (sql change): support option to compress output files using
gzip

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45978)
<!-- Reviewable:end -->

C0rWin commented:
Currently still work in progress

C0rWin committed:
importccl support option to compress output files using gzip

This commit extends EXPORT functionality by enabling compression of the
exported stream as suggested in #45579. Currently only gzip is supported
and the export clause to use compression looks as following:

```
export into csv 's3://export.csv' with compression = gzip from select * from foo;
```

Signed-off-by: Artem Barger <bartem@il.ibm.com>

Release note (sql change): support option to compress output files using
gzip

Release justification: none

dt commented on pkg/ccl/importccl/exportcsv.go:
```diff
@@ -31,6 +33,88 @@ import (
+}
+
+// ResetBuffer
+func (c *csvExporter) ResetBuffer() {
```
I think this should also be resetting the compressor, right?

C0rWin replied:
indeed

C0rWin replied:
After digging a bit into `gzip.Writer.Reset()` it doesn't writes anything into buffer while resetting it, just reset state of the gzip writer to its initial state. Therefore, it seems that either order will be fine, while I do agree that resetting buffer first would be more logically sound.

dt commented on pkg/sql/execinfrapb/processors_bulk_io.proto:
```diff
@@ -145,6 +143,17 @@ message CSVWriterSpec {
+  }
+
+  // compression code type to be used while epxporting CSV file,
+  // if it's null (i.e. not defined), no compression will be used.
```
```suggestion
```
The second line of the comment isn't needed: nullable = false actually means it _cannot_ be null (the field type will be `int` not `*int`) so there is no "null" or "undefined" state to reason about, just its the zero value which in this case is None.

dt commented on pkg/sql/execinfrapb/processors_bulk_io.proto:
```diff
@@ -145,6 +143,17 @@ message CSVWriterSpec {
+  // Compression list of the compression codecs which are currently
+  // supported for CSVWriter spec
+  enum Compression {
+    none = 0;
```
We usually capitalize the enum value names so the generated code looks more go-like.

C0rWin replied:
done

dt commented on pkg/sql/execinfrapb/processors_bulk_io.proto:
```diff
@@ -145,6 +143,17 @@ message CSVWriterSpec {
+
+  // Compression list of the compression codecs which are currently
+  // supported for CSVWriter spec
+  enum Compression {
```
I think "FileCompression" is enough of an independent concept we could move it top-level instead of nesting in CSV spec.

C0rWin replied:
sure

dt commented on pkg/sql/export.go:
```diff
@@ -117,10 +125,21 @@ func (ef *execFactory) ConstructExport(
 
+	// Check whenever compression is expected and extract compression codec name in case
+	// of positive result
+	codecName := ""
```
IMO  could just go straight to the enum here and skip the string.
```
var codec execinfrapb.FileCompression
if ... {
  codec = execinfrapb.FileCompression_Gzip
}
```

C0rWin replied:
done

dt commented on pkg/sql/distsql_physical_planner.go:
```diff
@@ -3219,11 +3219,19 @@ func (dsp *DistSQLPlanner) createPlanForExport(
+	// read value of compression codec, default is no compression
+	// otherwise lookup from supported set of codec values
+	compessionCodec := execinfrapb.CSVWriterSpec_none
+	if codec, exist := execinfrapb.CSVWriterSpec_Compression_value[n.codecName]; exist {
```
eh, i'm wary of this map lookup. mentioned below, but i'd rather we just use the enum type end-to-end, converting directly to it where to do the initial string validation and avoid this conversion step here.

C0rWin replied:
done

dt reviewed (commented):
This looks great! Thanks for picking it up!

A few comments inline.

dt commented:
As far as testing, i think you can just duplicate one of the existing tests, for example: 
https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/importccl/exportcsv_test.go#L181
but just confirm that a) the EXPORT runs with your option and b) you then open a file named "n1.0.csv*.gz*" and that that file decompresses to the expected content via a gzip stream.

C0rWin commented on pkg/ccl/importccl/exportcsv.go:
```diff
@@ -31,6 +33,88 @@ import (
+
+// Flush
+func (c *csvExporter) Flush() {
+	c.csvWriter.Flush()
```
Yep, it's necessary since while closing compressor also appends to archive headers.

C0rWin replied:
But to be explicit, I think it's better to call Close() with appropriate comment section to explain what has been done there, wdyt?

dt replied:
sounds go, but I think that means this method is unused now that the code below calls `Close`? If so, let's remove it.

C0rWin replied:
It's being used at line 236.

C0rWin commented:
> As far as testing, i think you can just duplicate one of the existing tests, for example:
https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/importccl/exportcsv_test.go#L181
but just confirm that a) the EXPORT runs with your option and b) you then open a file named "n1.0.csv*.gz*" and that that file decompresses to the expected content via a gzip stream.


thanks will add UT based on that example.

dt commented on pkg/ccl/importccl/exportcsv.go:
```diff
@@ -31,6 +32,98 @@ import (
+	}
+}
+
+// ResetBuffer
```
```suggestion
// ResetBuffer resets the buffer and compressor state.
```

C0rWin replied:
done

dt reviewed (approved):
This looks great! Just a few final nits inline.

I think you can ignore the red CI for the moment: the testrace failure is known flake (#45919) and the lint failure is just enforcing the 2 week code freeze.

C0rWin commented on pkg/ccl/importccl/exportcsv.go:
```diff
@@ -160,12 +245,16 @@ func (sp *csvWriter) Run(ctx context.Context) {
 			defer es.Close()
 
-			size := buf.Len()
+			size := writer.Len()
```
good point, yes, it has to be after close to count on footers as well.

C0rWin commented on pkg/ccl/importccl/exportcsv.go:
```diff
@@ -160,12 +245,16 @@ func (sp *csvWriter) Run(ctx context.Context) {
-			filename := strings.Replace(pattern, exportFilePatternPart, part, -1)
-			if err := es.WriteFile(ctx, filename, bytes.NewReader(buf.Bytes())); err != nil {
+			filename := writer.FileName(sp.spec, part)
+			// we need to close the writer here before we store content written so far
```
sure

C0rWin commented on pkg/ccl/importccl/exportcsv.go:
```diff
@@ -160,12 +245,16 @@ func (sp *csvWriter) Run(ctx context.Context) {
+			// we need to close the writer here before we store content written so far
+			// to the external storage, cause it also appends archive footer to the content
+			// in case compression was enabled
+			writer.Close()
```
👍

C0rWin commented on pkg/sql/export.go:
```diff
@@ -117,10 +126,21 @@ func (ef *execFactory) ConstructExport(
+	var codec execinfrapb.FileCompression
+	if name, ok := optVals[exportOptionCompression]; ok && len(name) != 0 {
+		if !strings.EqualFold(name, exportCompressionCodec) {
+			return nil, pgerror.New(pgcode.InvalidParameterValue, fmt.Sprintf("unsupported compression codec %s", name))
```
done

C0rWin commented on pkg/sql/execinfrapb/processors_bulk_io.proto:
```diff
@@ -145,6 +150,9 @@ message CSVWriterSpec {
   // chunk_rows is num rows to write per file. 0 = no limit.
   optional int64 chunk_rows = 4 [(gogoproto.nullable) = false];
+
+  // compression code type to be used while epxporting CSV file,
```
done

C0rWin commented:
>I  think you can ignore the red CI for the moment: the testrace failure is known flake (#45919) and the lint failure is just enforcing the 2 week code freeze.

understood and make sense 👍

dt commented:
Thank for taking this on!

Code looks good to go once master re-opens, though it looks like you might need to `make` again and amend to check in the generated proto diffs.

While you're at it, if you were so inclined, could you amend the commit message so the initial line has a colon after the package, and so it mentions EXPORT e.g. `importccl: support option to compress EXPORT files using gzip`. 

Other than that though, I think this can merge once we re-open master in the next week or two to new features. Thanks again!

C0rWin commented:
it seems that I've mistakenly forced push from master rather than development branch, which causes to auto-close pull-request. at any rate that wasn't intentional, pushed alternative one here: 46405

dt commented:
@C0rWin with the 20.1 branch cut today, we're about ready to re-open master to merging feature PRs. If you are some inclined, want to give this a quick rebase (and run `make` to ensure the protos are re-generated since it looks like there are some diffs there right now) and we can merge it?

C0rWin commented:
> @C0rWin with the 20.1 branch cut today, we're about ready to re-open master to merging feature PRs. If you are some inclined, want to give this a quick rebase (and run `make` to ensure the protos are re-generated since it looks like there are some diffs there right now) and we can merge it?

let me rebase and try.

C0rWin commented:
@dt running `make` leaves me with generated 

```
        modified:   pkg/sql/execinfrapb/processors_bulk_io.pb.go
        modified:   pkg/storage/enginepb/mvcc3.pb.go
        modified:   pkg/util/hlc/timestamp.pb.go
```

also seems after rebasing UT fails in TeamCity due to some issue with protos.

dt commented:
yeah, I think you want to check in those diffs as well -- a change to one proto can change all the generated ones and you just need to check in the generated diff as well. The CI is failing in the check that re-running generation produces no diffs.

C0rWin commented:
Unit test fails due to:


```
panic: event RetriableErr{CanAutoRetry:false, IsCommit:false} inappropriate in current state NoTxn{} [recovered]
	panic: panic while executing 1 statements: UPDATE _._ SET _ = $1 WHERE _ = $2; caused by event RetriableErr{CanAutoRetry:false, IsCommit:false} inappropriate in current state NoTxn{}
```

test in failure `TestBackupRestoreWithConcurrentWrites`, wonder how and whenever it actually related to current PR? fyi @dt

dt commented:
Whoops, sorry, the `TestBackupRestoreWithConcurrentWrites` failure is an unrelated flake recently introduced and fixed yesterday in #46829 -- might need one more rebase.

dt commented:
Oh, nevermind, looks like it passed!

Thanks again for sticking with this though a rough couple weeks in our dev cycle.

bors r=dt
